src/flint/types/fmpz_mod_poly.pyx

```diff
@@ -0,0 +1,1522 @@
+
+    def __rshift__(self, n):
+        return self.shift(-n)
+
+    @staticmethod
+    def _mod_(left, right):
+        cdef fmpz_t f
+        cdef fmpz_mod_poly res
+
+        # Case when left and right are already fmpz_mod_poly
+        if typecheck(left, fmpz_mod_poly) and typecheck(right, fmpz_mod_poly):
+            if not (<fmpz_mod_poly>left).ctx == (<fmpz_mod_poly>right).ctx:
+                raise ValueError("moduli must match")
+
+        # Case when right is not fmpz_mod_poly, try to convert to fmpz
+        elif typecheck(left, fmpz_mod_poly):
+            right = (<fmpz_mod_poly>left).ctx.any_as_fmpz_mod_poly(right)
+            if right is NotImplemented:
+                return NotImplemented
+
+        # Case when left is not fmpz_mod_poly, try to convert to fmpz
+        else:
+            left = (<fmpz_mod_poly>right).ctx.any_as_fmpz_mod_poly(left)
+            if left is NotImplemented:
+                return NotImplemented
+
+        res = fmpz_mod_poly.__new__(fmpz_mod_poly)
+        res.ctx = (<fmpz_mod_poly>left).ctx
+        fmpz_init(f)
+        fmpz_mod_poly_rem_f(
+            f, res.val, (<fmpz_mod_poly>left).val, (<fmpz_mod_poly>right).val, res.ctx.mod.val
+        )
+        if not fmpz_is_one(f):
+            fmpz_clear(f)
+            raise ValueError(
+                f"Cannot compute remainder of {left} modulo {right}"
+            )
+
+        return res
+
+    def __mod__(s, t):
+        return fmpz_mod_poly._mod_(s, t)
+
+    def __rmod__(s, t):
+        return fmpz_mod_poly._mod_(t, s)    
+
+    def __richcmp__(self, other, int op):
+        cdef bint res
+        if op != 2 and op != 3:
+            raise TypeError("fmpz_mod_poly cannot be ordered")
+
+        if not typecheck(other, fmpz_mod_poly):
+            other = self.ctx.any_as_fmpz_mod_poly(other)
+
+        if typecheck(other, fmpz_mod_poly):
+            res = (self.ctx == (<fmpz_mod_poly>other).ctx) and \
+                  fmpz_mod_poly_equal(self.val, (<fmpz_mod_poly>other).val, self.ctx.mod.val)
+            if op == 2:
+                return res
+            else:
+                return not res
+        else:
+            return NotImplemented
+
+    def __getitem__(self, long i):
+        cdef fmpz_mod x
+        x = fmpz_mod.__new__(fmpz_mod)
+        x.ctx = self.ctx.mod
+        if i < 0:
+            return x
+        fmpz_mod_poly_get_coeff_fmpz(
+            x.val, self.val, i, self.ctx.mod.val
+        )
+        return x
+
+    def __setitem__(self, long i, x):
+        if i < 0:
+            raise ValueError("cannot assign to index < 0 of polynomial")
+        v = self.ctx.mod.any_as_fmpz_mod(x)
+        if v is NotImplemented:
+            raise TypeError
+        fmpz_mod_poly_set_coeff_fmpz(
+            self.val, i, (<fmpz_mod>v).val, self.ctx.mod.val
+        )
+
+    def __len__(self):
+        return fmpz_mod_poly_length(self.val, self.ctx.mod.val)
+
+    def __hash__(self):
+        return hash(tuple(self.coeffs()))
```

Comment:
We need to decide what to do with this in general. Mutable objects should not define `__hash__`. The question is whether for polynomials mutability is more useful than hashing.

Reply:
I feel like hashing is more useful? I don't really know to be honest. I think other polynomial types have `__setitem__`, so maybe the right thing to do is remove the hash?